### ISSUE TITLE
Querying Vertices using properties while running an Update causes queries to fail. (Only Tested on Cosmos)

### ISSUE THREAD

**Issue by Euan-McVie** [ExRam/ExRam.Gremlinq#1092]
**Describe the bug**
When Running a query like:
```cs
_g
    .V<Portfolio>()
    .Where(portfolio => portfolio.Name == portfolio.Name)
```
And at the same time running an upsert that is not changing the value of property being queried using:
```cs
        _g
            .V<TVertex>(vertex.Id)
            .Where(v => v.TenantId == vertex.TenantId)
            .Fold()
            .Coalesce(__ => __
                .Unfold()
                .Update(vertex)
                .Property(x => x.Ttl, vertex.Ttl),
                __ => __.AddV(vertex))
```
The query will temporarily fail to find the record due to `Update` dropping all properties in the side effect.

There are other symptoms with the same root cause e.g. projections in the query result in `ServerEvaluationError`s being returned (we have applied a retry policy to mitigate)

**Expected behavior**
Less inconsistent querying behaviour during an Update.

I understand that this is at least partially related to how Cosmos is treating the drop side effect as not part of the same update operation as the Property steps that add the properties back in.

Also aware that the drop is required for `list` properties as otherwise the update will add extra values to the list instead of replacing it.

Preference would be for the `Update` to not drop single properties as that solves the issue for those. In our case we rarely use list properties.
e.g. 
```cs
            var droppableKeys = props
                .Where(p => p.value is not Traversal && p.value is IEnumerable && !environment.SupportsType(p.value.GetType()))
                .Select(static p => p.key)
                .ToArray();
```

Is there a reason that non-list properties require to be dropped during an update because of a provider other than cosmos's limitation?

Thanks

**Version information**
```xml
<PackageReference Include="ExRam.Gremlinq.Core" Version="11.0.0" />
<PackageReference Include="ExRam.Gremlinq.Providers.CosmosDb" Version="11.0.0" />
```

**Minimal working example**
TODO - I will copy out my workaround extension + test and add it as a PR to `test/Core.Tests/Issues/Github/` to show the issue and the code I used to apply the mitigation.
Note that as the issue is timing related the test I'm using to demonstrate is potentially flaky so this is nothing more than a demonstration of the issue rather than a test that would be worth considering for a real merge.

**In any case:**

- The repro code must be able to checkout and run in the matter of minutes. If the issue contains neither of these things and/or just incomplete code snippets it may be labelled accordingly and become stale.
Please understand that I ask for this much effort from you in exchange for my work since it is tedious and time-consuming to isolate the relevant bits from an issue each time just to setup
a new project that might not expose the unexpected behavior.

**Additional context**
- I understand that `MergeV` should solve this, but Cosmos does not support that yet so not going to request that addition to this library yet.
- I have not checked the Edge equivalent as we don't have non-cosmos properties on our edges at the moment.

**Comment by Euan-McVie** [comment]
Hi @danielcweber 
Good catch on nulls.

My version of `droppableKeys` from your `AddOrUpdate` is now:
```cs
static bool IsListValue(object value, IGremlinQueryEnvironment environment)
                => value is IEnumerable && value is not Traversal && !environment.SupportsType(value.GetType());

            var droppableKeys = props
                .Where(
                    p => p.value is null || IsListValue(p.value, environment))
                .Select(static p => p.key)
                .ToArray();
```
This works for us as we have very few lists (and none that currently have the transactional issue that cosmos has)

Would it be possible to add the equivalent to your real library so that when v12 is released we can drop our custom code? (and the copies of some of the internal cache and serialization stuff as it is internal to your lib)

Thanks
Euan

**Comment by danielcweber** [comment]
I implemented a different behaviour in version 12.0.0-preview.775. Properties will only be dropped first if the new value is either a list or null. Otherwise, a subsequent `.property()` will overwrite the value anyway. Feedback appredicated!

**Comment by Euan-McVie** [comment]
@danielcweber - I think you are good. 
My tests continue to pass after multiple iterations so it appears that having sideeffects in the middle does not stop an atomic update of properties either side of it. Very weird that having the sideeffect first doesn't work the same, but as it seems to work I'm not going to question further and not going to ask MS as time spent chasing through their support is probably not worth it.

Closing as resolved via #1111

**Comment by danielcweber** [comment]
Thanks for your feedback. I think I'd still refine this at some point to bundle the side effects and the properties after each other, like it used to be, but this time without unnecessary drops.
